sycl/source/detail/host_pipe.cpp

```diff
@@ -0,0 +1,70 @@
+//==-------------------- host_pipe.cpp -----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <detail/host_pipe_map_entry.hpp>
+#include <detail/program_manager/program_manager.hpp>
+#include <sycl/ext/intel/experimental/host_pipes.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+template <class _name, class _dataT, typename _propertiesT>
+_dataT host_pipe<_name, _dataT, _propertiesT>::read(queue &q,
+                                                    memory_order order) {
```

Comment:
This won't work. The templates need to be resolvable at compile-time, which they won't be if they are in the runtime library source files. Instead you can have a templateless base class for `host_pipe` and define the parts that need library-source-only features here, e.g. the `ProgramManager` instance.

Reply:
Thanks Steffen. I have fixed the code accordingly.